hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/WriteOperationHelper.java

```diff
@@ -257,17 +257,25 @@ public PutObjectRequest createPutObjectRequest(String destKey,
    * Create a {@link PutObjectRequest} request to upload a file.
    * @param dest key to PUT to.
    * @param sourceFile source file
+   * @param headers optional map of custom headers.
    * @return the request
    */
   @Retries.OnceRaw
-  public PutObjectRequest createPutObjectRequest(String dest,
-      File sourceFile) {
+  public PutObjectRequest createPutObjectRequest(
+      String dest,
+      File sourceFile,
+      @Nullable final Map<String, String> headers) {
```

Comment:
Why nullable? Empty map should be fine ?

Reply:
makes it easier to hand down in various places without having to repeatedly create empty files.